src/hotspot/share/opto/output.cpp

```diff
@@ -1862,6 +1862,19 @@ void PhaseOutput::fill_buffer(CodeBuffer* cb, uint* blk_starts) {
   if (C->print_assembly()) {
     ttyLocker ttyl;  // keep the following output all in one block
     if (!VMThread::should_terminate()) {  // test this under the tty lock
+      // print_metadata and dump_asm may safepoint which makes us loose the ttylock.
+      // We call them first and write to a stringStream, then we retake the lock to
+      // make sure the end tag is coherent, and that xmlStream->pop_tag is done thread safe.
+      ResourceMark rm;
+      stringStream method_metadata_str;
+      if (C->method() != NULL) {
+        C->method()->print_metadata(&method_metadata_str);
+      }
+      stringStream dump_asm_str;
+      dump_asm_on(&dump_asm_str, node_offsets, node_offset_limit);
```

Comment:
Does dump_asm_on also require InVM like print_metadata?  If it's not, is it easier to resume ttylock right after print_metadata?

Reply:
There is a `GUARDED_VM_ENTRY` that may safepoint in `dump_asm_on` -> `dump_on` -> `dump2` -> `print_name_on` -> `print_symbol_on`.